contracts/protocol/bases/PriceDiscoveryBase.sol

```diff
@@ -62,124 +72,250 @@ contract PriceDiscoveryBase is ProtocolBase {
+     * - Reseller did not approve protocol to transfer exchange token in escrow
+     * - New voucher owner is not buyer wallet
      *
-     * @param _exchangeId - the id of the exchange to commit to
-     * @param _exchangeToken - the address of the ERC20 token used for the exchange (zero address for native)
+     * @param _tokenId - the id of the token
+     * @param _offer - the fully populated BosonTypes.Offer struct
      * @param _priceDiscovery - the fully populated BosonTypes.PriceDiscovery struct
-     * @param _initialSellerId - the id of the original seller
      * @return actualPrice - the actual price of the order
      */
     function fulfilBidOrder(
-        uint256 _exchangeId,
-        address _exchangeToken,
+        uint256 _tokenId,
+        Offer storage _offer,
         PriceDiscovery calldata _priceDiscovery,
-        uint256 _initialSellerId
+        address owner,
+        ProtocolLib.ProtocolLookups storage _lookups,
+        IBosonVoucher bosonVoucher
     ) internal returns (uint256 actualPrice) {
-        IBosonVoucher bosonVoucher = IBosonVoucher(protocolLookups().cloneAddress[_initialSellerId]);
+        address priceDiscoveryContract = _lookups.priceDiscoveryContractByVoucher[_tokenId];
+
+        // If caller is not the owner, check if the owner exists and is voucher price discovery contract
+        require(owner != address(0) && owner == priceDiscoveryContract, OWNER_MUST_BE_PRICE_DISCOVERY_CONTRACT);
+
+        address exchangeToken = _offer.exchangeToken;
+
+        // Native token is not safe, as its value can be intercepted in the receive function.
+        if (exchangeToken == address(0)) exchangeToken = address(weth);
+
+        // check the last owner's balance since they are the one who should receive the price via the price discovery contract
+        address lastOwner = _lookups.lastVoucherOwner[_tokenId];
+
+        require(lastOwner != address(0), LAST_OWNER_NOT_FOUND);
+
+        // Check last owner balance before the exchange
+        uint256 balanceBefore = getBalance(exchangeToken, lastOwner);
+
+        // Track native balance just in case if seller send some native currency or price discovery contract does
+        uint256 protocolNativeBalanceBefore = getBalance(address(0), owner);
+
+        // Store the information about incoming voucher
+        ProtocolLib.ProtocolStatus storage ps = protocolStatus();
+
+        {
+            address cloneAddress = _lookups.cloneAddress[_offer.sellerId];
+
+            callPriceDiscoveryContract(_priceDiscovery, cloneAddress);
+
+            uint256 balanceAfter = getBalance(exchangeToken, lastOwner);
+
+            actualPrice = balanceAfter - balanceBefore;
+
+            require(actualPrice >= _priceDiscovery.price, INSUFFICIENT_VALUE_RECEIVED);
+
+            // Transfer funds to protocol - caller must pay on behalf of the seller
+            FundsLib.validateIncomingPayment(exchangeToken, actualPrice);
```

Comment:
Another thing, since `exchangeToken` is forced to be WETH for offers in the native token, the seller is here forced to wrap ethers before, which shouldn't be necessary in this case.

Reply:
This cannot be made optional because we can only assume the true price after finishing the call to the price discovery contract.  Seller funds are encumbered inside the call to the PD contract, and we can't assume the price as the protocol balance difference at this point because we can't assume that PD will first transfer the exchange token to the token owner (seller) and only then transfer the voucher to the buyer.